File: apps/agent/tests/integration/list_keys_test.go
package integration_test

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/require"
	"github.com/unkeyed/unkey/apps/agent/pkg/integration"
	"github.com/unkeyed/unkey/apps/agent/pkg/server"
)

func TestListKeys(t *testing.T) {
	createApiResponse := integration.Step[server.CreateApiResponse]{
		Name:   "Create API",
		Method: "POST",
		Url:    fmt.Sprintf("%s/v1/apis.createApi", BASE_URL),
		Header: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
		},
		Body: map[string]any{
			"name": "scenario-test-pls-delete",
		},
	}.Run(t)
	require.Equal(t, 200, createApiResponse.Status)
	require.NotEmpty(t, createApiResponse.Body.ApiId)
	require.NotEmpty(t, createApiResponse.Header.Get("Unkey-Trace-Id"))

	defer func() {

		deleteApiresponse := integration.Step[map[string]any]{
			Name:   "Delete API",
			Method: "POST",
			Url:    fmt.Sprintf("%s/v1/apis.deleteApi", BASE_URL),
			Header: map[string]string{
				"Content-Type":  "application/json",
				"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
			},
			Body: map[string]any{
				"apiId": createApiResponse.Body.ApiId,
			},
		}.Run(t)
		require.Equal(t, 200, deleteApiresponse.Status)
		require.NotEmpty(t, deleteApiresponse.Header.Get("Unkey-Trace-Id"))
	}()

	// Create 5 keys
	for i := 0; i < 5; i++ {
		createKeyResponse := integration.Step[server.CreateKeyResponse]{
			Name:   "Create Key",
			Method: "POST",
			Url:    fmt.Sprintf("%s/v1/keys.createKey", BASE_URL),
			Header: map[string]string{
				"Content-Type":  "application/json",
				"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
			},
			Body: map[string]any{
				"apiId": createApiResponse.Body.ApiId,
			},
		}.Run(t)
		require.Equal(t, 200, createKeyResponse.Status)
		require.NotEmpty(t, createKeyResponse.Body.Key)
		require.NotEmpty(t, createKeyResponse.Body.KeyId)
		require.NotEmpty(t, createKeyResponse.Header.Get("Unkey-Trace-Id"))

		defer func() {

			revokeKeyResponse := integration.Step[map[string]any]{
				Name:   "Revoke Key",
				Method: "POST",
				Url:    fmt.Sprintf("%s/v1/keys.deleteKey", BASE_URL),
				Header: map[string]string{
					"Content-Type":  "application/json",
					"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
				},
				Body: map[string]any{
					"keyId": createKeyResponse.Body.KeyId,
				},
			}.Run(t)
			require.Equal(t, 200, revokeKeyResponse.Status)
			require.NotEmpty(t, revokeKeyResponse.Header.Get("Unkey-Trace-Id"))
		}()

	}

	listKeysResponse := integration.Step[server.ListKeysResponse]{
		Name:   "List Keys",
		Method: "GET",
		Url:    fmt.Sprintf("%s/v1/apis/%s/keys", BASE_URL, createApiResponse.Body.ApiId),
		Header: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("Bearer %s", ROOT_KEY),
		},
	}.Run(t)
	require.Equal(t, 200, listKeysResponse.Status)
	require.NotEmpty(t, listKeysResponse.Header.Get("Unkey-Trace-Id"))
	require.Equal(t, 5, len(listKeysResponse.Body.Keys))

}

File: apps/api/src/integration/list_keys.test.ts
import { IntegrationHarness } from "@/pkg/testutil/integration-harness";
import type { V1ApisCreateApiRequest, V1ApisCreateApiResponse } from "@/routes/v1_apis_createApi";
import type { V1ApisDeleteApiRequest, V1ApisDeleteApiResponse } from "@/routes/v1_apis_deleteApi";
import type { V1ApisListKeysResponse } from "@/routes/v1_apis_listKeys";
import type { V1KeysCreateKeyRequest, V1KeysCreateKeyResponse } from "@/routes/v1_keys_createKey";
import type { V1KeysDeleteKeyRequest } from "@/routes/v1_keys_deleteKey";
import { expect, test } from "vitest";

test(
  "create and list keys",
  async (t) => {
    const h = await IntegrationHarness.init(t);
    const { key: rootKey } = await h.createRootKey(["*"]);

    const createApiResponse = await h.post<V1ApisCreateApiRequest, V1ApisCreateApiResponse>({
      url: `${h.baseUrl}/v1/apis.createApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        name: "scenario-test-pls-delete",
      },
    });
    expect(createApiResponse.status).toEqual(200);
    expect(createApiResponse.body.apiId).toBeDefined();
    expect(createApiResponse.headers).toHaveProperty("unkey-request-id");

    for (let i = 0; i < 5; i++) {
      const createKeyResponse = await h.post<V1KeysCreateKeyRequest, V1KeysCreateKeyResponse>({
        url: `${h.baseUrl}/v1/keys.createKey`,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${rootKey}`,
        },
        body: {
          apiId: createApiResponse.body.apiId,
          byteLength: 32,
          prefix: "test",
          enabled: true,
        },
      });
      expect(createKeyResponse.status).toEqual(200);
      expect(createKeyResponse.body.keyId).toBeDefined();
      expect(createKeyResponse.body.key).toBeDefined();
    }
    const listKeysResponse = await h.get<V1ApisListKeysResponse>({
      url: `${h.baseUrl}/v1/apis.listKeys?apiId=${createApiResponse.body.apiId}`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
    });

    expect(listKeysResponse.status).toEqual(200);
    expect(listKeysResponse.body.keys).toHaveLength(5);

    /**
     * Teardown
     */
    const deleteApi = await h.post<V1ApisDeleteApiRequest, V1ApisDeleteApiResponse>({
      url: `${h.baseUrl}/v1/apis.deleteApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
      },
    });
    expect(deleteApi.status, `status mismatch - ${JSON.stringify(deleteApi)}`).toEqual(200);
  },
  { timeout: 30_000 },
);

test(
  "with revalidate",
  async (t) => {
    const h = await IntegrationHarness.init(t);
    const { key: rootKey } = await h.createRootKey(["*"]);

    const createApiResponse = await h.post<V1ApisCreateApiRequest, V1ApisCreateApiResponse>({
      url: `${h.baseUrl}/v1/apis.createApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        name: "scenario-test-pls-delete",
      },
    });
    expect(createApiResponse.status).toEqual(200);
    expect(createApiResponse.body.apiId).toBeDefined();
    expect(createApiResponse.headers).toHaveProperty("unkey-request-id");

    for (let i = 1; i <= 10; i++) {
      const createKeyResponse = await h.post<V1KeysCreateKeyRequest, V1KeysCreateKeyResponse>({
        url: `${h.baseUrl}/v1/keys.createKey`,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${rootKey}`,
        },
        body: {
          apiId: createApiResponse.body.apiId,
          byteLength: 32,
          prefix: "test",
          enabled: true,
        },
      });
      expect(createKeyResponse.status).toEqual(200);
      expect(createKeyResponse.body.keyId).toBeDefined();
      expect(createKeyResponse.body.key).toBeDefined();
      const listKeysResponse = await h.get<V1ApisListKeysResponse>({
        url: `${h.baseUrl}/v1/apis.listKeys?apiId=${createApiResponse.body.apiId}&revalidateKeysCache=true`,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${rootKey}`,
        },
      });

      expect(listKeysResponse.status).toEqual(200);
      expect(listKeysResponse.body.keys).toHaveLength(i);
    }

    /**
     * Teardown
     */
    const deleteApi = await h.post<V1ApisDeleteApiRequest, V1ApisDeleteApiResponse>({
      url: `${h.baseUrl}/v1/apis.deleteApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
      },
    });
    expect(deleteApi.status, `status mismatch - ${JSON.stringify(deleteApi)}`).toEqual(200);
  },
  { timeout: 30_000 },
);

test(
  "list keys does not return revoked keys",
  async (t) => {
    const h = await IntegrationHarness.init(t);
    const { key: rootKey } = await h.createRootKey(["*"]);
    const createApiResponse = await h.post<V1ApisCreateApiRequest, V1ApisCreateApiResponse>({
      url: `${h.baseUrl}/v1/apis.createApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        name: "scenario-test-pls-delete",
      },
    });
    expect(
      createApiResponse.status,
      `status mismatch - ${JSON.stringify(createApiResponse)}`,
    ).toEqual(200);
    expect(createApiResponse.body.apiId).toBeDefined();
    expect(createApiResponse.headers).toHaveProperty("unkey-request-id");

    const activeKeyIds: string[] = [];
    const deletedKeyIds: string[] = [];

    for (let i = 0; i < 10; i++) {
      const createKeyResponse = await h.post<V1KeysCreateKeyRequest, V1KeysCreateKeyResponse>({
        url: `${h.baseUrl}/v1/keys.createKey`,
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${rootKey}`,
        },
        body: {
          apiId: createApiResponse.body.apiId,
          byteLength: 32,
          prefix: "test",
          enabled: true,
        },
      });
      expect(createKeyResponse.status).toEqual(200);
      expect(createKeyResponse.body.keyId).toBeDefined();
      expect(createKeyResponse.body.key).toBeDefined();

      if (i % 2 === 0) {
        const deleteKeyResponse = await h.post<V1KeysDeleteKeyRequest, V1KeysDeleteKeyRequest>({
          url: `${h.baseUrl}/v1/keys.deleteKey`,
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${rootKey}`,
          },
          body: {
            keyId: createKeyResponse.body.keyId,
          },
        });
        expect(
          deleteKeyResponse.status,
          `status mismatch - ${JSON.stringify(deleteKeyResponse)}`,
        ).toEqual(200);
        deletedKeyIds.push(createKeyResponse.body.keyId);
      } else {
        activeKeyIds.push(createKeyResponse.body.keyId);
      }
    }
    const listKeysResponse = await h.get<V1ApisListKeysResponse>({
      url: `${h.baseUrl}/v1/apis.listKeys?apiId=${createApiResponse.body.apiId}`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
    });

    expect(listKeysResponse.status).toEqual(200);
    expect(listKeysResponse.body.keys).toHaveLength(5);
    for (const keyId of activeKeyIds) {
      expect(listKeysResponse.body.keys.map((k) => k.id)).toContain(keyId);
    }
    for (const keyId of deletedKeyIds) {
      expect(listKeysResponse.body.keys.map((k) => k.id)).not.toContain(keyId);
    }

    /**
     * Teardown
     */
    const deleteApi = await h.post<V1ApisDeleteApiRequest, V1ApisDeleteApiResponse>({
      url: `${h.baseUrl}/v1/apis.deleteApi`,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${rootKey}`,
      },
      body: {
        apiId: createApiResponse.body.apiId,
      },
    });
    expect(deleteApi.status, `status mismatch - ${JSON.stringify(deleteApi)}`).toEqual(200);
  },
  { timeout: 30_000 },
);
